### ISSUE TITLE
allclose() vector `atol` parameter wrong shape if NaN in a or b.

### ISSUE THREAD

**Issue by pavelkomarov** [numpy/numpy#14320]
```python
>>>np.allclose([1,2,float('nan')], [1,2,float('nan')], equal_nan=True)
True
>>>np.allclose([1,2,float('nan')], [1,2,float('nan')], atol=[1,1], equal_nan=True)
True
>>>np.allclose([1,2,3], [1,2,3)], atol=[1,1,1], equal_nan=True)
True
>>>np.allclose([1,2,float('nan')], [1,2,float('nan')], atol=[1,1,1], equal_nan=True)
ValueError: operands could not be broadcast together with shapes (3,) (2,)
```

It appears that `allclose` is basically removing entries where `a` or `b` is NaN before passing them on to `isclose` with the same `atol` provided to `allclose`. This of course doesn't work if I expect there to be no NaNs some of the time and need to provide a longer `atol` because my `a` and `b` are different scales in different positions.

What should happen is `allclose` removes the same positions from `atol` before passing it on, so the dimensions always agree.

**Comment by eric-wieser** [comment]
I'm not sure support for a vector in atol was intended, but adding support would be fine by me

**Comment by maxwell-aladago** [comment]
> I'm not sure support for a vector in atol was intended, but adding support would be fine by me

The unintended support for vectors came from broadcasting inside [`within_tol`](https://github.com/numpy/numpy/blob/ccfbcc1cd9a4035a467f2e982a565ab27de25b6b/numpy/core/numeric.py#L2352) in `np.isclose()`. 
 
As a support for vector `atol` and `rtol`, will the following addition below the [line](https://github.com/numpy/numpy/blob/ccfbcc1cd9a4035a467f2e982a565ab27de25b6b/numpy/core/numeric.py#L2376) be correct @eric-wieser?
```
atol = full(x.shape, atol)
rtol = full(x.shape, rtol)
```
These two lines will raise `ValueError: could not broadcast input array from shape (p) into shape (q)` if `atol` or `rtol` is not a scaler and cannot also be casted to the same shape as `x`.  If desired,  the error can be caught and re-raised with a more informative message like `'atol must be a scalar or of shape %s' %x.shape`.

The line 
```
 cond[finite] = within_tol(x[finite], y[finite], atol, rtol)
```
 can now then be written as 
```
 cond[finite] = within_tol(x[finite], y[finite], atol[finite], rtol[finite])
```

If `atol` and `rtol` must strictly be scalars, then I will suggest that a check is added to make that more explicit.

**Comment by mattip** [comment]
I am not sure I understand the use case for non-scalar `rtol`/`atol`. I closed #14343, and I think a PR to make sure they are scalars would be more helpful since that is what is in the docs, with perhaps a deprecation cycle.

**Comment by pavelkomarov** [comment]
A nonscalar `atol` makes sense if you want to ensure position 1 across two vectors is within `a` of each other, position 2 is within `b` of each other, etc., in only a single call.
